KEP-4222: Add roundtrip tests to Unstructured via CBOR and JSON.

benluddy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature
/sig api-machinery

#### What this PR does / why we need it:

This is an alpha requirement for KEP-4222 (CBOR). It tests the paths native-to-CBOR-to-unstructured-to-CBOR-to-native and native-to-JSON-to-unstructured-to-JSON-to-native, verifying that the original object is roundtripped faithfully and that the intermediate unstructured objects are identical across the CBOR and JSON paths. This is especially relevant for the case of a dynamic client interacting with built-in API resources using the CBOR encoding.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/4222
```

benluddy commented:
/test pull-kubernetes-unit

deads2k commented on staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go:
```diff
@@ -683,6 +684,16 @@ func (q Quantity) MarshalJSON() ([]byte, error) {
 }
 
+// todo
+func (q Quantity) MarshalCBOR() ([]byte, error) {
```
unit test for these and others.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+
+var em cbor.EncMode = func() cbor.EncMode {
+	em, err := cbor.EncOptions{
+		Sort:          cbor.SortBytewiseLexical,
```
some of these aren't obvious.  Comment them to explain what cases these are covering for us?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+
+type defaultMetaFactory struct{}
+
+var mfdm cbor.DecMode = func() cbor.DecMode {
```
Why should we have a different set of decoding options for pulling metadata?  I'd expect a single set.

Any response should be a comment here because I bet future me will wonder the same thing.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+	return &actual, nil
+}
+
+type Serializer struct {
```
type assertion

Also, can this be private and only constructed via `NewSerializer` returning the interface?  Serializer + RecognizingDecoder?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+}
+
+// gvkWithDefaults returns group kind and version defaulting from provided default
+func gvkWithDefaults(actual, defaultGVK schema.GroupVersionKind) schema.GroupVersionKind {
```
reading this, I regret ever allowing gvk to be hinted.  I wonder how much code breaks if we change that on decoding.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+	"github.com/fxamacker/cbor/v2"
+)
+
+var em cbor.EncMode = func() cbor.EncMode {
```
nit: extra letters don't cost extra, expand the names for the abbreviations.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+
+			*actual = into.GetObjectKind().GroupVersionKind()
+			if len(actual.Kind) == 0 {
+				return nil, actual, runtime.NewMissingKindErr("<cbor>") // todo: diagnostic notation
```
flagging so these don't merge without some kind of error.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -0,0 +1,224 @@
+		types, _, err := s.typer.ObjectKinds(into)
+		switch {
+		case runtime.IsNotRegisteredError(err):
+			if err := dm.Unmarshal(data, into); err != nil {
```
huh, so if the type is not Unstructured and is not registered to the scheme, we try to decode into it anyway?  Seems like we should be able to fail.

benluddy commented:
/triage accepted
/remove-sig architecture
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig network
/remove-sig node
/remove-sig storage
/remove-sig testing

benluddy commented:
/remove-sig architecture
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig network
/remove-sig node
/remove-sig storage
/remove-sig testing

benluddy commented:
/test pull-kubernetes-unit

benluddy commented:
/remove-sig architecture
/remove-sig auth
/remove-sig cli
/remove-sig cloud-provider
/remove-sig cluster-lifecycle
/remove-sig instrumentation
/remove-sig network
/remove-sig node
/remove-sig storage

benluddy committed:
Implement cbor.Marshaler and cbor.Unmarshaler for IntOrString.

benluddy committed:
Implement cbor.Marshaler and cbor.Unmarshaler for metav1.Time.

benluddy committed:
Implement cbor.Marshaler and cbor.Unmarshaler for metav1.MicroTime.

benluddy committed:
Implement cbor.Marshaler and cbor.Unmarshaler for resource.Quantity.

benluddy committed:
Fuzz roundtrip to Unstructured via both JSON and CBOR.

benluddy commented:
/cc @stlaz

benluddy committed:
Update indirect dependencies with ./hack/update-vendor.sh.

Implementing custom marshaling on several API types for CBOR makes the upstream CBOR library an
indirect dependency of several staging modules.

benluddy commented:
/retest-required

stlaz commented on pkg/api/testing/unstructured_test.go:
```diff
@@ -134,6 +142,208 @@ func TestRoundTrip(t *testing.T) {
+			{Version: "v1", Kind: "RangeAllocation"},
+			{Version: "v1", Kind: "ConfigMap"},
+			{Version: "v1", Kind: "ConfigMapList"},
+			{Group: "admissionregistration.k8s.io", Version: "v1beta1", Kind: "MutatingWebhookConfiguration"},
```
Should APIService be here, too? https://github.com/kubernetes/kubernetes/blob/68091805a53c6fdd42553f8ccdbf932ab263fbdd/staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/types.go#L69
Perhaps it's not a part of the legacyscheme, but might be worth noting?

benluddy replied:
Yes, the aggregator and apiextensions types should be covered, too. I'll make sure this is included on the checklist.

stlaz commented on pkg/api/testing/unstructured_test.go:
```diff
@@ -134,6 +142,208 @@ func TestRoundTrip(t *testing.T) {
+
+				// original->JSON/CBOR->Unstructured == original->JSON/CBOR->Unstructured->CBOR->Unstructured
+				buf.Reset()
+				if err := cborSerializer.Encode(item, &buf); err != nil {
```
should `item` be `uJSON` or `uCBOR`?

benluddy replied:
Great catch, thanks! Yes, it should be one of those.

benluddy commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -324,3 +327,215 @@ func TestParse(t *testing.T) {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if diff := cmp.Diff(got, tc.want); diff != "" {
+				t.Errorf("unexpected difference between expected and actual output:\n%s", diff)
```
No, IIRC I tested this and diff is smart enough to format byte slices legibly using hex.

benluddy commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -324,3 +327,215 @@ func TestParse(t *testing.T) {
+		assertOnError func(*testing.T, error)
+	}{
+		{
+			in: []byte{0xa0}, // {}
```
I don't think it's necessary to test input kinds exhaustively here. Those cases will all have coverage in specific unmarshal test cases. For this test it's enough (IMO) to have one case that covers a non-nil error on unmarshaling into int32.

benluddy commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time_test.go:
```diff
@@ -147,6 +150,60 @@ func TestTimeMarshalJSONUnmarshalYAML(t *testing.T) {
+		in   Time
+		out  []byte
+	}{
+		{name: "zero value", in: Time{}, out: []byte{0xf6}},                                                                                        // null
```
Added, thanks.

stlaz commented:
/lgtm

deads2k commented on pkg/api/testing/unstructured_test.go:
```diff
@@ -134,6 +142,208 @@ func TestRoundTrip(t *testing.T) {
+// objects produced by both encodings must be identical and be themselves roundtrippable to JSON and
+// CBOR.
+func TestRoundtripToUnstructured(t *testing.T) {
+	// These are GVKs that whose CBOR roundtrippability is blocked by a known issue that must be
+	// resolved as a prerequisite for alpha.
```
have you broken this out in the tracking issue?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/api/resource/quantity.go:
```diff
@@ -683,6 +685,12 @@ func (q Quantity) MarshalJSON() ([]byte, error) {
+func (q Quantity) MarshalCBOR() ([]byte, error) {
+	// The call to String() should never return the string "<nil>" because the receiver's
+	// address will never be nil.
+	return cbor.Marshal(q.String())
```
explained.  It's more costly to make everything json and then cbor and back again. The fuzz tests ensure that we're faithfully converting.

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/micro_time.go:
```diff
@@ -129,6 +131,25 @@ func (t *MicroTime) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
+func (t *MicroTime) UnmarshalCBOR(b []byte) error {
```
can I get a fuzz test?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/time.go:
```diff
@@ -116,6 +118,25 @@ func (t *Time) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
+func (t *Time) UnmarshalCBOR(b []byte) error {
```
can I get a fuzz test?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr.go:
```diff
@@ -126,6 +141,17 @@ func (intstr IntOrString) MarshalJSON() ([]byte, error) {
 	}
 }
 
+func (intstr IntOrString) MarshalCBOR() ([]byte, error) {
```
can I get a fuzz test?

deads2k commented on staging/src/k8s.io/apimachinery/pkg/util/intstr/intstr_test.go:
```diff
@@ -324,3 +330,239 @@ func TestParse(t *testing.T) {
+	}
+}
+
+func TestIntOrStringRoundtripCBOR(t *testing.T) {
```
somewhat surprised this one works.  I expected to see a custom fuzzer that decides to set the int part or the string part instead of potentially using both.  Does this work reliably?

benluddy replied:
It implements fuzz.Interface in `intstr_fuzz.go` (similar story for Time and MicroTime).

deads2k replied:
ah

benluddy replied:
Also, I took a look at the inputs it was generating to make sure these methods were actually being used:

```
=== RUN   TestIntOrStringRoundtripCBOR
    intstr_test.go:551: {0 1128564365 }
    intstr_test.go:551: {0 200166354 }
    intstr_test.go:551: {0 -276099788 }
    intstr_test.go:551: {0 -1249228784 }
    intstr_test.go:551: {0 1742569458 }
    intstr_test.go:551: {0 -1375958113 }
    intstr_test.go:551: {1 0 (jʢǔ腔誖4Þ[{謴ǡ嶆Jɏ嗑}
    intstr_test.go:551: {0 1906691861 }
    intstr_test.go:551: {1 0 TĆ帎ȼCǲ屁ɻEwû与塩衰噺瀷Ȋ椮}
    intstr_test.go:551: {0 1459274389 }
    intstr_test.go:551: {1 0 }
    intstr_test.go:551: {0 -467127729 }
    intstr_test.go:551: {0 -705730334 }
    intstr_test.go:551: {1 0 溿ŷʈ沙t:励¬Ⱦ}
    intstr_test.go:551: {1 0 鷩}
    intstr_test.go:551: {0 -126567898 }
    intstr_test.go:551: {0 953288002 }
    intstr_test.go:551: {0 -1326681651 }
    intstr_test.go:551: {1 0 謳嬅Ɵ<炋靼瓽M蜢ɣȷ}
    intstr_test.go:551: {1 0 *堗ƛ強ɵpŵŴȭ-ģ2}
    intstr_test.go:551: {0 -1009384221 }
    intstr_test.go:551: {1 0 ɖ昮ǀǝ&ǻ}ń窚X`'整}
    intstr_test.go:551: {0 1993804638 }
    intstr_test.go:551: {0 2066970653 }
    intstr_test.go:551: {1 0 筄D垞}
    intstr_test.go:551: {0 -1084349919 }
    intstr_test.go:551: {0 1653629984 }
    intstr_test.go:551: {0 784751765 }
    intstr_test.go:551: {0 620224278 }
    intstr_test.go:551: {1 0 uÑē靎ĽIɭÓ=謋Ƀ喠靬?Ǜ}
    intstr_test.go:551: {0 577636246 }
    intstr_test.go:551: {0 -650130851 }
    intstr_test.go:551: {1 0 @Ś4v}
```

deads2k commented:
testing looks good and thorough.  this is the next step for CBOR serialization and leaves other serializers alone.

/lgtm
/approve

liggitt commented:
/approve
for go.mod bumps
